pupil-api/src/services/log.service.ts

```diff
@@ -0,0 +1,108 @@
+import * as winston from 'winston'
+import config from '../config'
+
+/*
+const syslogLevels = {
+  emerg: 0,
+  alert: 1,
+  crit: 2,
+  error: 3,
+  warning: 4,
+  notice: 5,
+  info: 6,
+  debug: 7
+}
+*/
+
+export class Logger {
+  private level: string
+  private logger
+
+  constructor () {
+    this.level = config.Logging.LogLevel
+
+    let format
+    if (config.Logging.ApplicationInsights.LogToWinston) {
+      format = winston.format.simple()
+    } else {
+      format = winston.format.combine(
+        winston.format.colorize({ all: true }),
+        winston.format.simple()
+      )
+    }
+
+    const baseLogOptions = {
+      levels: winston.config.syslog.levels,
+      level: this.level,
+      format: format,
+      transports: [
+        new winston.transports.Console({ level: this.level, silent: false, consoleWarnLevels: ['warn', 'error'] })
+      ],
+      meta: true,
+      expressFormat: true,
+      colorize: false
+    }
+
+    this.logger = winston.createLogger(baseLogOptions)
+
+    if (config.Logging.ApplicationInsights.LogToWinston) {
+      console.log('app insights config enabled')
+      const { AzureApplicationInsightsLogger } = require('winston-azure-application-insights')
+      const appInsights = require('applicationinsights')
+
+      appInsights.setup(config.Logging.ApplicationInsights.Key).start()
+
+      this.logger
+        .clear() // remove all transports
+        .add(new AzureApplicationInsightsLogger({
+          insights: appInsights,
+          sendErrorsAsExceptions: true,
+          level: this.level
+        }))
+    }
+  }
+
+  log (level, msg, exception) {
+    this.logger.log(level, msg, exception)
+  }
+
+  /**
+   * AI -> critical
+   * @param {string} msg
+   */
+  alert (msg, exception = null) { this.log('alert', msg, exception) }
+
+  /**
+   * AI -> error
+   * @param {string} msg
+   */
+  error (msg, exception = null) { this.log('error', msg, exception) }
+
+  /**
+   * AI -> warning
+   * @param {string} msg
+   */
+  warn (msg, exception = null) { this.log('warning', msg, exception) }
+
+  /**
+   * AI -> notice
+   * @param {string} msg
+   */
+  info (msg, exception = null) { this.log('info', msg, exception) }
+
+  /**
+   * AI -> verbose
+   * @param {string} msg
+   */
+  debug (msg, exception = null) { this.log('debug', msg, exception) }
+
+  /**
+   * Return the underlying `winston` logger
+   * @return {winston.Logger | *}
+   */
+  getLogger () {
+    return this.logger
+  }
+}
+
+export default new Logger()
```

Comment:
My reading of this line is that it will generate a new logger every time it is  `require`ed.   In my opinion, the logger should be a singleton.

What do you think?

Reply:
Seems wrong to me as well, good spot @jon-shipley

Reply:
Seems wrong to me as well, good spot @jon-shipley

Reply:
Heh, it _looks_ like it does it every time it's required, but really that's the singleton pattern with TS - it instantiates the class and then exports it.

It might depend on the require cache, but really it's the same as the admin app approach.